vendors/st/boards/stm32l475_discovery/ports/common_io/iot_spi.c

```diff
@@ -0,0 +1,679 @@
+                           uint8_t * const pvBuffer,
+                           size_t xBytes )
+{
+    int32_t lError = IOT_SPI_SUCCESS;
+
+    /* TODO: How to find a way to tell if slave is not selected */
+    if( ( pxSPIPeripheral == NULL ) || ( pxSPIPeripheral->sOpened == IOT_SPI_CLOSED ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else if( ( pvBuffer == NULL ) || ( xBytes == 0 ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else
+    {
+        SPI_HandleTypeDef * pxSpi = pxSPIPeripheral->pxSpiContext->pxSpi;
+
+        if( HAL_SPI_GetState( pxSpi ) == HAL_SPI_STATE_BUSY_RX )
+        {
+            lError = IOT_SPI_BUS_BUSY;
+        }
+        else if( HAL_SPI_Receive( pxSpi, pvBuffer, ( uint16_t ) xBytes, IOT_SPI_BLOCKING_TIMEOUT ) != HAL_OK )
+        {
+            lError = IOT_SPI_TRANSFER_ERROR;
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_spi_read_async( IotSPIHandle_t const pxSPIPeripheral,
+                            uint8_t * const pvBuffer,
+                            size_t xBytes )
+{
+    int32_t lError = IOT_SPI_SUCCESS;
+
+    /* TODO: Add case for slave not selected? */
+    if( ( pxSPIPeripheral == NULL ) || ( pxSPIPeripheral->sOpened == IOT_SPI_CLOSED ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else if( ( pvBuffer == NULL ) || ( xBytes == 0 ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else
+    {
+        SPI_HandleTypeDef * pxSpi = pxSPIPeripheral->pxSpiContext->pxSpi;
+
+        if( HAL_SPI_GetState( pxSpi ) == HAL_SPI_STATE_BUSY_RX )
+        {
+            lError = IOT_SPI_BUS_BUSY;
+        }
+        else
+        {
+            HAL_NVIC_SetPriority( ( IRQn_Type ) ( ( STM32_SPI_HalContext_t * ) pxSPIPeripheral->pxSpiContext )->eIrqNum, 1, 0 );
+            HAL_NVIC_EnableIRQ( ( IRQn_Type ) ( ( STM32_SPI_HalContext_t * ) pxSPIPeripheral->pxSpiContext )->eIrqNum );
+
+            if( HAL_SPI_Receive_IT( pxSpi, pvBuffer, ( uint16_t ) xBytes ) != HAL_OK )
+            {
+                lError = IOT_SPI_TRANSFER_ERROR;
+            }
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_spi_write_sync( IotSPIHandle_t const pxSPIPeripheral,
+                            uint8_t * const pvBuffer,
+                            size_t xBytes )
+{
+    int32_t lError = IOT_SPI_SUCCESS;
+
+    /* TODO: Add case for slave not selected? */
+    if( ( pxSPIPeripheral == NULL ) || ( pxSPIPeripheral->sOpened == IOT_SPI_CLOSED ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else if( ( pvBuffer == NULL ) || ( xBytes == 0 ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else
+    {
+        SPI_HandleTypeDef * pxSpi = pxSPIPeripheral->pxSpiContext->pxSpi;
+
+        if( HAL_SPI_GetState( pxSpi ) == HAL_SPI_STATE_BUSY_TX )
```

Comment:
Since it's synchronous, should it try to block until the state is no longer busy?

Reply:
This question made me think. I believe that is not the design goal of Common IO. We leave this decision to the upper layer.